rfcs/rdm-0080-bulk-importer.md

```diff
@@ -0,0 +1,104 @@
+The resources an instance has could be extended using entrypoints.
+The **Serializer** will be responsible for reading the input file and transforming each row according to a set of "embedded rules". In the future it would be nice to also have a dump option to 1) download the file that was uploaded if edit were made and 2) generate an empty file  with all the possible fields as a template.
+As for the resources, these can be discovered/extended using entrypoints.
+
+**Note on custom fields**:
+Each serializer implementation is responsible for reading their particular section within `IMPORTER_CUSTOM_FIELDS`. As an initial approach it can look something like this:
+```python
+IMPORTER_CUSTOM_FIELDS={
+	'csv_rdm_record_serializer':[
+		{
+			'columns': ['a', 'b', 'c'],
+			'transformer': transformer_func
+		}
+	]
+}
+```
+The signature of the transform function should be same of the `transform` function of the base class, i.e. taking the current row being transformed and current object where transformations are stored.
+The columns function will be use to validate the input file, i.e. check that CSV headers are correct.
+NOTE: No custom field would be consider mandatory, any column listed here can be missing from the input file.
+
+![](0080/uml_2.png)
+
+### JSON Schemas
+Both, the Importer and the ImporterRecord have a json field that will have a schema to enforce some kind of structure.
+#### Importer
+- `status`: can possibly be computed from records
+- `mode`:
+	- Possible values: `import`, `delete`
+- `title`
+- `description` / `notes`
+- `resource`
+- `serializer`
+- `start_time`
+- `end_time`
+#### ImporterRecord
+
+- `status`
+	- Possible values: `pending`, `running`, `success`, `failure`, `stopped`, `archived` (`warning`?)
+- `message`
+- `info`: key:value array to be displayed on the UI in the form of columns.
+- `src`: string/text representation of the raw data coming from the input file
+- `transformed`: (dict) result of the load call from the serializer
+### Administration interface
+
+![](0080/mockup.jpg)
+
+## Drawbacks
+- Representing nested fields (such us author's metadata) inside a flat file, like CVS, is not optimal. 
+## Alternatives
+There is some work schedule to happen beginig of April to add to expand the current InvenioRDM REST APIs, to work with "arrays/streams" of "typed" actions, similar to how the [ElasticSearch Bulk API](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html) works. There's nothing on paper yet, but it will be our opportunity to explore with a "power-user" to see if this is feasible and useful.
+Maybe the importer endpoint could only deal with the load and transform part, and hand over the rest to the Bulk API.
+## Unresolved questions
+
+- Metadata file re-upload to correct errors
+- How do we set a file for preview?
```

Comment:
is there any limit on the import size or maximum number of files? It might be quite troublesome to handle big imports

Reply:
Since I wrote this, we have shown it to potential users, and this has come up on several occasions.
I think we will have the same "limitations" as the current deposit form. To mitigate this, we are pondering allowing users to enter known URIs into the files column. Imagine a shared location the service would have access to, like a bucket on AWS/GCP or directly a URL that can be fetched somehow.

Reply:
the problem might be both on the CSV upload and associated file upload. The CSV /marcxml itself will be problematic if it contains a lot of records, for 2 reasons: size of the csv itself, and the time the underlying task will take to process it - depending on how much memory celery worker has available

Reply:
You're definitely right. We did consider the problem with the celery workers' memory. Unfortunately, I don't have a solution other than making the process as memory-efficient as possible. 
For example, we plan to start a task for each record (row) inside the input file and let that task do the transformation and validation so we will process one record at a time.
I think adding an "artificial" limit to the number of records at this point might not make a lot of sense. Once we have the process in place, and knowing where it can "break", we can load test it and set an informative limit.